java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,123 @@
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+import arcs.core.data.Claim
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.ParticleSpec
+import arcs.core.data.Recipe
+import arcs.core.data.expression.Expression
+
+/**
+ * Deduce [Claim]s on Paxel [Expression]s from all connections on a [Recipe.Particle].
+ *
+ * See [DependencyNode] for more information.
+ */
+fun RecipeGraph.Node.Particle.deduceClaims(): List<Claim> {
+  return particle.spec.connections.values
+    .filter { connection -> connection.expression != null }
+    .flatMap { connection ->
+      val root = connection.expression?.analyze()
+      require(root is DependencyNode.AssociationNode) {
```

Comment:
Just curious: what will happen if `root` is `null`? Will this fail?

Reply:
Yes, it should fail if the root is null.